merkle/merkle.go

```diff
@@ -106,35 +108,72 @@ func MerkleCreate(baseSize uint64, seed Trytes, offset uint64, security Security
 
 	treeMerkleSize := MerkleSize(baseSize)
 	tree := make(Trits, treeMerkleSize*HashTrinarySize)
+	treeLock := &sync.RWMutex{}
 
 	h := sponge.GetSpongeFunc(spongeFunc, curl.NewCurlP27)
 
 	td := MerkleDepth(treeMerkleSize) - 1
 
-	// create base addresses
-	for leafIndex := uint64(0); leafIndex < baseSize; leafIndex++ {
-		subSeed, err := signing.Subseed(seed, offset+MerkleLeafIndex(leafIndex, baseSize), h)
-		if err != nil {
-			return nil, err
-		}
-
-		key, err := signing.Key(subSeed, security, h)
-		if err != nil {
-			return nil, err
-		}
+	waitGroup := &sync.WaitGroup{}
+	errorChan := make(chan error)
+	waitChan := make(chan struct{})
+
+	numCPU := uint64(runtime.NumCPU())
+	leafesPerCPU := baseSize / numCPU
+
+	var i uint64
+	for i = 0; i < numCPU; i++ {
+		waitGroup.Add(1)
+
+		go func(i uint64, spongeFunc sponge.SpongeFunction) {
+			h := spongeFunc.Clone()
+
+			// create base addresses
+			for leafIndex := i * leafesPerCPU; leafIndex < (i+1)*leafesPerCPU; leafIndex++ {
+				subSeed, err := signing.Subseed(seed, offset+MerkleLeafIndex(leafIndex, baseSize), h)
+				if err != nil {
+					errorChan <- err
+					return
+				}
+
+				key, err := signing.Key(subSeed, security, h)
+				if err != nil {
+					errorChan <- err
+					return
+				}
+
+				keyDigests, err := signing.Digests(key, h)
+				if err != nil {
+					errorChan <- err
+					return
+				}
+
+				address, err := signing.Address(keyDigests, h)
+				if err != nil {
+					errorChan <- err
+					return
+				}
+
+				treeIdx := MerkleNodeIndex(td, leafIndex, td)
+				treeLock.Lock()
+				copy(tree[treeIdx*HashTrinarySize:(treeIdx+1)*HashTrinarySize], address)
+				treeLock.Unlock()
+			}
 
-		keyDigests, err := signing.Digests(key, h)
-		if err != nil {
-			return nil, err
-		}
+			waitGroup.Done()
+		}(i, h)
+	}
 
-		address, err := signing.Address(keyDigests, h)
-		if err != nil {
-			return nil, err
-		}
+	// Create signal when finished
+	go func() {
+		waitGroup.Wait()
+		waitChan <- struct{}{}
+	}()
 
-		treeIdx := MerkleNodeIndex(td, leafIndex, td)
-		copy(tree[treeIdx*HashTrinarySize:(treeIdx+1)*HashTrinarySize], address)
+	select {
```

Comment:
This is actually schematically wrong. If any of the goroutine already has produced an error, then all other goroutines will be stuck forever trying to send into the error channel. Consider collecting the errors happening within the Merke tree generation and put them into a slice of errors (protected by a mutex of course). You can also drop the `waitChan` and just use the normal `WaitGroup` pattern. After you `wait()`ed, you can check if the error slice contains any errors.